core/src/ledger/ibc/mod.rs

```diff
@@ -136,27 +134,47 @@ where
     /// Store the denom when transfer with MsgRecvPacket
     fn store_denom(&mut self, envelope: MsgEnvelope) -> Result<(), Error> {
         match envelope {
-            MsgEnvelope::Packet(PacketMsg::Recv(msg)) => {
-                let data = match serde_json::from_slice::<PacketData>(
-                    &msg.packet.data,
-                ) {
-                    Ok(data) => data,
-                    // not token transfer data
-                    Err(_) => return Ok(()),
-                };
-                let prefix = TracePrefix::new(
-                    msg.packet.port_id_on_b.clone(),
-                    msg.packet.chan_id_on_b,
-                );
-                let mut coin = data.token;
-                coin.denom.add_trace_prefix(prefix);
-                let trace_hash = storage::calc_hash(coin.denom.to_string());
-                self.ctx
-                    .borrow_mut()
-                    .store_denom(trace_hash, coin.denom)
-                    .map_err(|e| {
-                        Error::Denom(format!("Write the denom failed: {}", e))
-                    })
+            MsgEnvelope::Packet(PacketMsg::Recv(_)) => {
+                let result = self
+                    .ctx
+                    .borrow()
+                    .get_ibc_event("denomination_trace")
+                    .map_err(|_| {
+                        Error::Denom("Reading the IBC event failed".to_string())
+                    })?;
+                if let Some(event) = result {
+                    if let (Some(trace_hash), Some(ibc_denom)) = (
+                        event.attributes.get("trace_hash"),
+                        event.attributes.get("denom"),
```

Comment:
to avoid writing two chained ifs, you can map and flatten the options
https://doc.rust-lang.org/std/option/enum.Option.html#method.flatten